core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java

```diff
@@ -149,22 +159,24 @@ public UnderFileSystem addMount(long mountId, String ufsUri, Map<String, String>
       throws IOException {
     Preconditions.checkArgument(mountId != IdUtils.INVALID_MOUNT_ID, "mountId");
     Preconditions.checkArgument(ufsUri != null, "uri");
-    UnderFileSystem ufs = getOrAdd(ufsUri, ufsConf);
-    mMountIdToUnderFileSystemMap.put(mountId, ufs);
-    return ufs;
+    Pair<UnderFileSystem, AtomicLong> ufsCounter = getOrAdd(ufsUri, ufsConf);
+    mMountIdToUnderFileSystemMap.put(mountId, ufsCounter);
+    return ufsCounter.getFirst();
   }
 
   @Override
   public void removeMount(long mountId) {
     Preconditions.checkArgument(mountId != IdUtils.INVALID_MOUNT_ID, "mountId");
-    // TODO(binfan): check the refcount of this ufs in mUnderFileSystemMap and remove it if this is
-    // no more used. Currently, it is possibly used by out mount too.
-    mMountIdToUnderFileSystemMap.remove(mountId);
+    Pair<UnderFileSystem, AtomicLong> ufsCounter = mMountIdToUnderFileSystemMap.get(mountId);
+    // Remove ufs from mMountIdToUnderFileSystemMap if its counter comes to zero.
+    if (ufsCounter.getSecond().decrementAndGet() == 0L) {
+      mMountIdToUnderFileSystemMap.remove(mountId);
```

Comment:
ufs is already registered in closer `mCloser.register(ufsCounter.getFirst());` just want to make sure if we should explicitly call close().